rfcs/20200712-tfrt-kernel-fallback.md

```diff
@@ -0,0 +1,647 @@
+
+Currently, [TF Lite](https://www.tensorflow.org/lite) supports a
+[limited set of ops](https://www.tensorflow.org/lite/guide/ops_compatibility).
+As the range and variety of applications grows, it becomes essential to grow the
+pool of available ops as well, ideally supporting everything that fully-fledged
+TensorFlow supports now.
+
+However, supporting TensorFlow ops on mobile devices presents some challenges.
+Specifically, binary size on mobile platforms should be restricted. TF Lite team
+provided us with the following *ideal* numbers:
+
+*   100-200k overhead to call TF kernels
+*   20k / kernel marginal size
+
+To get closer to the size restrictions we plan to define a call path from TFRT
+to TensorFlow kernels that minimizes the amount of generated code.
+
+## User Benefit
+
+Running more kernels on mobile devices would allow TensorFlow Lite users to implement
+a wider range of models.
+
+## Design Proposal
+
+We propose to call the kernel’s Compute method directly from
+[TFRT](https://github.com/tensorflow/runtime) without going through TensorFlow
+Eager C API first. We introduce kernel context and registration implementation that support core
+kernel functionality with minimal dependencies.
+
+High-level diagram of the proposed design:
+
+<img src="https://github.com/annarev/community/blob/tfrt_kernel_fallback_rfc/rfcs/20200712-tfrt-kernel-fallback/KernelFallbackDiagram.png" alt="Kernel Fallback design diagram." width="400px">
+
+## Kernel registration
+
+We will use a separate registry for kernels supported by TFRT forwarding. To do
+so, we will define a `TFRTOpKernelFactories` class that would keep a map from
+kernel name to a list of registrations.
+
+```cpp
+class TFRTOpKernelFactories {
+ public:
+  TFRTOpKernelFactories();
+  void RegisterFactory(StringPiece kernel_class_name,
+                       TFRTOpKernelReg kernel_info);
+
+  // Creates a kernel with the given name and passes op_kernel_construction
+  // to kernel constructor.
+  // Returns the constructed kernel on success.
+  // In case of failure, returns a nullptr. Kernel creation can fail in one
+  // of the following cases:
+  //   1. Kernel with the given name is not found.
+  //   2. Attributes in op_kernel_construction don't match type constraints
+  //      for any of the kernels with this name.
+  //      Note that we consider a constraint to be "not matched" if attribute
+  //      it applies to is not in op_kernel_construction.
+  std::unique_ptr<TFRTOpKernel> CreateKernel(
+      StringPiece kernel_class_name,
+      TFRTOpKernelConstruction* op_kernel_construction) const;
+
+ private:
+  llvm::StringMap<std::vector<TFRTOpKernelReg>> factories_;
+};
+
+extern llvm::ManagedStatic<TFRTOpKernelFactories>
+    tfrt_forwarding_kernel_factories;
+```
+
+Similar to current TensorFlow kernel registartion, we will introduce a
+registration macro that adds a kernel to `TFRTOpKernelFactories`.
+
+```cpp
+#define REGISTER_KERNEL_FALLBACK_KERNEL(name, ...) \
```

Comment:
shall we spell this as REGISTER_FALLBACK_KERNEL for simplicity?

Reply:
+1

Reply:
Updated.